hastory-server/test/Data/Hastory/ServerSpec.hs

```diff
@@ -28,10 +30,16 @@ spec =
           Left (FailureResponse _ resp) <- runClientM req siClientEnv
           responseBody resp `shouldBe` "Invalid Token provided."
           responseStatusCode resp `shouldBe` status403
-    context "with correct token" $
+    context "with correct token" $ do
       it "saves entry to database" $ \ServerInfo {..} ->
         forAllValid $ \syncRequest -> do
           _ <- runClientM (appendCommand siToken syncRequest) siClientEnv
           let selectEntries = fmap SQL.entityVal <$> SQL.selectList [] []
           dbEntries <- SQL.runSqlPool selectEntries siPool
-          dbEntries `shouldBe` [SyncRequest.toServerEntry syncRequest]
+          dbEntries `shouldBe` [toServerEntry syncRequest]
+      it "is idempotent" $ \ServerInfo {..} ->
+        forAllValid $ \syncRequest -> do
+          replicateM_ 2 $ runClientM (appendCommand siToken syncRequest) siClientEnv
+          let selectEntries = fmap SQL.entityVal <$> SQL.selectList [] []
+          dbEntries <- SQL.runSqlPool selectEntries siPool
+          length (dbEntries :: [ServerEntry]) `shouldBe` 1
```

Comment:
This is a beautiful test. 
Instead of just checking that there is only one entry, maybe we should expect an exact entry ?

Reply:
I just thought about this a bit more. We probably want these as three separate tests:

1. Idempotency: the db looks the same after the first sync as after the second sync.
2. When you upload one entry twice, the db still only has one entry
3. When you upload an entry, you expect that exact entry in the db.

Reply:
The reasoning is about what happens when we start allowing batch uploads.
1. stays exactly the same and will "just work" for N entries.
2. will be adapted from 1 to N
3. will stay exactly the same: test with a batch of one.

Reply:
I'll be working on these tests tonight!